### PR TITLE
🚨🚨 Source Amplitude: migrate to YAML

### LINKED ISSUE #22142: Source Amplitude: Migrate connector from GA (Python) to GA (YAML)

### PR DESCRIPTION
**Pull request by artem1205**
## What
Resolving:

- #22142 
- #21010
- #19036
- #16943

Migrate to YAML

## How
*Describe the solution*

## Recommended reading order
1. `y.python`

## 🚨 User Impact 🚨
`stream state` and `cursor field` for  stream event has been changed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

### CODE REVIEW

**Review by lazebnyi — APPROVED** [review]

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      record_selector:
+        type: RecordSelector
+        extractor:
+          type: CustomRecordExtractor
+          field_pointer: [ "data" ]
+          class_name: source_amplitude.components.ActiveUsersRecordExtractor
+    $options:
+      name: "active_users"
+      primary_key: "date"
+      path: "/2/users"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:124]: I'm a little bit confused here. It seems like the [/2/users API ](https://www.docs.developers.amplitude.com/analytics/apis/dashboard-rest-api/#get-active-and-new-user-counts) returns something like:
```
{
    "data": {
        "series": [ 
            [46109, 47542],
            [42845, 42626]
        ],
        "seriesMeta": ["United States", "Canada"],
        "xValues": ["2017-08-14", "2017-08-15"]
    }
}
```

However in this manifest, it uses on the `ActiveUsersRecordExtractor` extractor [which expect `seriesLabel`](https://github.com/airbytehq/airbyte/pull/22362/files#diff-ddf43f76d82f3374fff8f08974773811a594ed89a753f52d2462988fd5e622a1R44) and I don't see `seriesLabel` in the template response from Amplitude. Is this expected?

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        return record
+
+    def extract_records(
+        self,
+        response: requests.Response,
+    ) -> List[Record]:
+        try:
+            zip_file = zipfile.ZipFile(io.BytesIO(response.content))
+        except zipfile.BadZipFile as e:
+            logger.exception(e)
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py:96]: Why both `logger.exception` and `logger.error`? Couldn't we just do this? 
```
            logger.exception(
                f"Received an invalid zip file in response to URL: {response.request.url}."
                f"The size of the response body is: {len(response.content)}"
            )
```

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml (hunk 1 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:124]: Yes, this is expected:
image

**Review by roman-yermilov-gl — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py (hunk 2 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _date_time_to_rfc3339(self, record: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        """
+        Transform 'date-time' items to RFC3339 format
+        """
+        date_time_fields = self._get_date_time_items_from_schema()
+        for item in record:
+            if item in date_time_fields:
+                dt_value = record[item]
+                if not dt_value:
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    def extract_records(self, response: requests.Response) -> List[Record]:
+        response_data = response.json().get("data", [])
+        if response_data:
+            # From the Amplitude documentation it follows that "series" is an array with one element which is itself
+            # an array that contains the average session length for each day.
+            # https://developers.amplitude.com/docs/dashboard-rest-api#returns-2
+            series = response_data.get("series", [])
+            if len(series) > 0:
+                series = series[0]  # get the nested list
+                return [{"date": date, "length": series[i]} for i, date in enumerate(response_data["xValues"])]
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        return []
+
+
+class ActiveUsersRecordExtractor(RecordExtractor):
+    def extract_records(self, response: requests.Response) -> List[Record]:
+        response_data = response.json().get("data", [])
+        if response_data:
+            series = list(map(list, zip(*response_data["series"])))
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py]: updated

**Review by roman-yermilov-gl — APPROVED** [review]

File: airbyte-integrations/connectors/source-amplitude/acceptance-test-config.yml [hunk 6]
```diff
@@ -1,24 +1,42 @@
...
+      - config_path: "secrets/config.json"
+        status: "succeed"
+      - config_path: "integration_tests/invalid_config.json"
+        status: "failed"
   discovery:
-    - config_path: "secrets/config.json"
+    tests:
+      - config_path: "secrets/config.json"
+        backward_compatibility_tests_config:
+          disable_for_version: 0.1.21
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-amplitude/acceptance-test-config.yml:19]: can you clarify why we're disabling backwards compatibility tests and how this impacts users in production?

File: airbyte-integrations/connectors/source-amplitude/setup.py [hunk 7]
```diff
@@ -6,10 +6,14 @@
 from setuptools import find_packages, setup
 
 MAIN_REQUIREMENTS = [
-    "airbyte-cdk~=0.1",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-amplitude/setup.py:9]: why remove this constraint?

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import requests
+from airbyte_cdk.sources.declarative.extractors.record_extractor import RecordExtractor
+from airbyte_cdk.sources.declarative.schema.json_file_schema_loader import JsonFileSchemaLoader
+from airbyte_cdk.sources.declarative.types import Config, Record
+from dataclasses_jsonschema import JsonSchemaMixin
+
+logger = logging.getLogger("airbyte")
+
+
+class AverageSessionLengthRecordExtractor(RecordExtractor):
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py:23]: can you create issues for all these custom components as described [here](https://docs.google.com/document/d/1LB4gBk0slHj3zNmq1Qtj1U7VkKyTN9eFsVldgQ8T4HQ/edit?usp=drivesdk)?

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/spec.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      airbyte_secret: true
+    start_date:
+      type: string
+      title: Replication Start Date
+      pattern: "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$"
+      description: UTC date and time in the format 2021-01-25T00:00:00Z. Any data
+        before this date will not be replicated.
+      examples:
+      - "2021-01-25T00:00:00Z"
+    request_time_range:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/spec.yaml:40]: can you clarify why this is being added as part of this PR? I would have thought a migration PR strives to keep exactly the same external UX

**Review by sherifnada — CHANGES_REQUESTED** [review]
please update this to use the Beta lowcode CDK syntax. See the message in #using-the-cdk from brian

File: airbyte-integrations/connectors/source-amplitude/setup.py (hunk 7 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/setup.py:9]: GA connectors should always be on latest version of airbyte-cdk

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/spec.yaml (hunk 9 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/spec.yaml:40]: I've added `request_time_range` as part of https://github.com/airbytehq/airbyte/issues/21010, this is optional field.

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 10]
```diff
@@ -0,0 +1,171 @@
+version: "0.1.0"
+
+definitions:
+  selector:
+    type: RecordSelector
+    extractor:
+      type: DpathExtractor
+      field_pointer:
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:8]: Using the latest version of the airbyte-cdk, this should be `field_path`

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        - "{{ options.get('data_field') }}"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:9]: Using the latest version of the airbyte-cdk, `options` has been renamed `parameters`

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+          action: IGNORE
+          error_message: No data collected
+        - http_codes: [ 504 ]
+          action: IGNORE
+          error_message: The amount of data is large causing a timeout. For large amounts of data, the Amazon S3 destination is recommended.
+
+  retriever:
+    type: SimpleRetriever
+    record_selector:
+      $ref: "*ref(definitions.selector)"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:34]: Using the latest version of the airbyte-cdk, `$ref` now have the format `#/definitions/selector`

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    retriever:
+      $ref: "*ref(definitions.retriever)"
+      requester:
+        $ref: "*ref(definitions.requester)"
+        request_options_provider:
+          request_parameters:
+            start: "{{format_datetime(stream_slice.start_time, '%Y%m%d') }}"
+            end: "{{format_datetime(stream_slice.end_time, '%Y%m%d') }}"
+      stream_slicer:
+        $ref: "*ref(definitions.date_stream_slicer)"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:82]: Using the latest version of the airbyte-cdk, this should be under `incremental_sync` which is directly under the stream

**Review by maxi297 — COMMENTED** [review]
Make sure this is updated to the latest version of airbyte-cdk

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml (hunk 12 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:34]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml (hunk 11 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:9]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml (hunk 10 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+      path: "/3/cohorts"
+      data_field: "cohorts"
+
+  base_incremental_stream:
+    checkpoint_interval: 10
+    retriever:
+      $ref: "*ref(definitions.retriever)"
+      requester:
+        $ref: "*ref(definitions.requester)"
+        request_options_provider:
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml (hunk 13 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml]: updated

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py (hunk 8 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/components.py:22]: updated

File: airbyte-integrations/connectors/source-amplitude/acceptance-test-config.yml (hunk 6 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/acceptance-test-config.yml]: stream state and cursor field for stream event has been changed.

File: airbyte-integrations/connectors/source-amplitude/setup.py (hunk 7 above)
Comment (sherifnada) [reply airbyte-integrations/connectors/source-amplitude/setup.py:9]: practically speaking `~=` allows you to get that effect but still ensures you don't accidentally upgrade to a new major version (which we would expect a breaking change)

Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/setup.py:9]: Yep, but I thought we have always keep our GA up-to-date on latest version CDK, am I right?
Everyday connectors build report helps us in this case.

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+  retriever:
+    type: SimpleRetriever
+    record_selector:
+      $ref: "#/definitions/selector"
+    paginator:
+      type: NoPagination
+    requester:
+      $ref: "#/definitions/requester"
+
+  date_stream_slicer:
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:39]: nit: should be `datetime_incremental_sync` or something that does not reflect the stream slices

**Review by maxi297 — APPROVED** [review]
Once Sherif's requested changes are implemented, I'm good with this PR

**Review by sherifnada — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+    http_method: "GET"
+    authenticator:
+      type: BasicHttpAuthenticator
+      username: "{{ config['api_key'] }}"
+      password: "{{ config['secret_key'] }}"
+    error_handler:
+      type: DefaultErrorHandler
+      response_filters:
+        - http_codes: [ 400 ]
+          action: IGNORE
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml]: You're right, ignore needed only for code 404.
updated.

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml (hunk 15 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml]: updated

**Review by sherifnada — CHANGES_REQUESTED** [review]
one blocking comment about incremental sync

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+    incremental_sync:
+      $ref: "#/definitions/datetime_incremental_sync"
+      step: "PT{{config.get('request_time_range', 24)}}H"
+      cursor_field: "{{ parameters.get('stream_cursor_field') }}"
+      cursor_granularity: PT1H
+      start_datetime:
+        datetime: "{{ format_datetime(config['start_date'], '%Y-%m-%dT%H:%M:%S.%f%z') }}"
+      end_datetime:
+        datetime: "{{ now_utc().strftime('%Y-%m-%dT%H:%M:%S.%f%z') }}"
+      datetime_format: "%Y-%m-%dT%H:%M:%S.%f%z"
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:147]: We use `YYYYMMDDTHH` format in [request_params](https://github.com/airbytehq/airbyte/pull/22362/files#diff-b3e449e4461a8914c0664b227fa9a1dd70127cb17a06d2e9402965b2b8d1835aR129), while the field in records stored in `'%Y-%m-%dT%H:%M:%S.%f%z`

File: airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+      path: "/2/sessions/average"
+      stream_cursor_field: "date"
+
+  active_users_stream:
+    $ref: "#/definitions/base_incremental_stream"
+    retriever:
+      $ref: "#/definitions/base_incremental_stream/retriever"
+      requester:
+        $ref: "#/definitions/base_incremental_stream/retriever/requester"
+        request_parameters:
```
[the remark this answers is not in the document]
Comment (artem1205) [reply airbyte-integrations/connectors/source-amplitude/source_amplitude/manifest.yaml:102]: we have different request date-time format in `events` stream.

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-amplitude/acceptance-test-docker.sh [hunk 19]
```diff
@@ -1,9 +1,9 @@
 #!/usr/bin/env sh
 
 # Build latest connector image
-docker build . -t $(cat acceptance-test-config.yml | grep "connector_image" | head -n 1 | cut -d: -f2)
+docker build . -t $(cat acceptance-test-config.yml | grep "connector_image" | head -n 1 | cut -d: -f2-)
 
-# Pull latest acctest image
+# Pull latest acceptance test image
```
Comment (girarda) [inline airbyte-integrations/connectors/source-amplitude/acceptance-test-docker.sh:6]: nit: Can we either not make this change or apply it to all connectors (in a separate PR) for the sake of consistency?

**Review by girarda — APPROVED** [review]

**Comment by jnr0790** [comment]
Can we hold off on this until I reach out to the 8 users on cloud using Amplitude? I suggest releasing Tuesday the 28th.

**Comment by jnr0790** [comment]
Do we have an ETA on when this will be released?

**Comment by jnr0790** [comment]
cc @artem1205 @erica-airbyte

**Pull request merged**